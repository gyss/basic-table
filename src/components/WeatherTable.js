/** @jsx jsx */

import React from 'react'
import {connect} from 'react-redux'
import {css, jsx} from '@emotion/core'

import {AutoSizer, Table, Column} from 'react-virtualized'

import {fetchWeather, selectWeatherItem} from '../actions'

const container = css`
  height: 80%;
  width: 100%;
`

const header = css`
  padding: calc(2em + 10px);
  background-color: #fafbfc;
  box-shadow: inset 0 -1px 0 #e1e4e8;
`

const table = css`
  overflow: auto;
  height: 100%;
  padding: 2em;
`

function rowClassName({index}) {
  if (index < 0) {
    return 'headerRow'
  } else {
    return index % 2 === 0 ? 'evenRow' : 'oddRow'
  }
}

const WeatherTable = ({dispatch, loading, error, city, list, selected}) => {
  React.useEffect(() => {
    dispatch(fetchWeather())
  }, [])

  if (loading) {
    return <div>Loading weather data...</div>
  }

  if (error) {
    return <div>ERROR!</div>
  }

  function handleRowClick(event) {
    dispatch(selectWeatherItem(event.rowData))
  }

  return (
    <div css={container}>
      <div css={header}>
        <h2>
          {city.name}, {city.country}
        </h2>
      </div>
      <div css={table}>
        <AutoSizer>
          {({height, width}) => (
            <Table
              width={Math.max(1570, width)}
              height={height}
              headerHeight={40}
              rowHeight={30}
              rowCount={list.length}
              rowGetter={({index}) => list[index]}
              rowClassName={rowClassName}
              onRowClick={handleRowClick}>
              <Column label="Date" dataKey="dt_txt" width={200} flexGrow={1} />
              <Column width={110} label="Visibility" dataKey="visibility" flexGrow={1} />
              <Column
                width={100}
                label="Wind"
                dataKey="wind"
                cellDataGetter={({rowData}) => rowData.wind.speed + ' / ' + rowData.wind.deg}
                flexGrow={1}
              />
              <Column
                width={100}
                label="Clouds"
                dataKey="clouds"
                cellDataGetter={({rowData}) => rowData.clouds.all}
                flexGrow={1}
              />
              <Column
                width={100}
                label="Rain"
                dataKey="rain"
                cellDataGetter={({rowData}) => rowData.rain && rowData.rain['3h']}
                flexGrow={1}
              />
              <Column
                width={120}
                label="Feels like"
                dataKey="feels_like"
                cellDataGetter={({rowData}) => rowData.main.feels_like}
                flexGrow={1}
              />
              <Column
                width={140}
                label="Grnd Level"
                dataKey="grnd_level"
                cellDataGetter={({rowData}) => rowData.main.grnd_level}
                flexGrow={1}
              />
              <Column
                width={120}
                label="Humidity"
                dataKey="humidity"
                cellDataGetter={({rowData}) => rowData.main.humidity}
                flexGrow={1}
              />
              <Column
                width={120}
                label="Pressure"
                dataKey="pressure"
                cellDataGetter={({rowData}) => rowData.main.pressure}
                flexGrow={1}
              />
              <Column
                width={120}
                label="Sea Level"
                dataKey="sea_level"
                cellDataGetter={({rowData}) => rowData.main.sea_level}
                flexGrow={1}
              />
              <Column
                width={120}
                label="Temp"
                dataKey="temp"
                cellDataGetter={({rowData}) => rowData.main.temp}
                flexGrow={1}
              />
              <Column
                width={120}
                label="Temp min"
                dataKey="temp_min"
                cellDataGetter={({rowData}) => rowData.main.temp_min}
                flexGrow={1}
              />
              <Column
                width={120}
                label="Temp max"
                dataKey="temp_max"
                cellDataGetter={({rowData}) => rowData.main.temp_max}
                flexGrow={1}
              />
            </Table>
          )}
        </AutoSizer>
      </div>
    </div>
  )
}

const mapStateToProps = (state) => ({
  city: state.weather.city,
  list: state.weather.list,
  loading: state.weather.loading,
  error: state.weather.error,
  selected: state.weather.selected,
})

export default connect(mapStateToProps)(WeatherTable)
