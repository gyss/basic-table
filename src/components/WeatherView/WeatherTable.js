/** @jsx jsx */

import React from 'react'
import PropTypes from 'prop-types'
import {css, jsx} from '@emotion/core'
import {AutoSizer, Table, Column} from 'react-virtualized'

import {kelvinToCelsius} from '../../utils/conversions'

function rowClassName({index}) {
  if (index < 0) {
    return 'headerRow'
  } else {
    return index % 2 === 0 ? 'evenRow' : 'oddRow'
  }
}

const WeatherTable = ({onSelect, weather}) => {
  return (
    <AutoSizer>
      {({height, width}) => (
        <Table
          width={Math.max(1570, width)}
          height={height}
          headerHeight={40}
          rowHeight={30}
          rowCount={weather.list.length}
          rowGetter={({index}) => weather.list[index]}
          rowClassName={rowClassName}
          onRowClick={onSelect}>
          <Column
            label="Date"
            dataKey="dt_txt"
            width={200}
            cellDataGetter={({rowData}) => rowData.dt_txt.substr(0, 16)}
            flexGrow={1}
          />
          <Column width={110} label="Visibility" dataKey="visibility" flexGrow={1} />
          <Column
            width={100}
            label="Wind"
            dataKey="wind"
            cellDataGetter={({rowData}) => rowData.wind.speed + ' / ' + rowData.wind.deg}
            flexGrow={1}
          />
          <Column
            width={100}
            label="Clouds"
            dataKey="clouds"
            cellDataGetter={({rowData}) => rowData.clouds.all}
            flexGrow={1}
          />
          <Column
            width={100}
            label="Rain"
            dataKey="rain"
            cellDataGetter={({rowData}) => rowData.rain && rowData.rain['3h']}
            flexGrow={1}
          />
          <Column
            width={140}
            label="Grnd Level"
            dataKey="grnd_level"
            cellDataGetter={({rowData}) => rowData.main.grnd_level}
            flexGrow={1}
          />
          <Column
            width={120}
            label="Humidity"
            dataKey="humidity"
            cellDataGetter={({rowData}) => rowData.main.humidity + ' %'}
            flexGrow={1}
          />
          <Column
            width={120}
            label="Pressure"
            dataKey="pressure"
            cellDataGetter={({rowData}) => rowData.main.pressure}
            flexGrow={1}
          />
          <Column
            width={120}
            label="Sea Level"
            dataKey="sea_level"
            cellDataGetter={({rowData}) => rowData.main.sea_level}
            flexGrow={1}
          />
          <Column
            width={120}
            label="Feels like"
            dataKey="feels_like"
            cellDataGetter={({rowData}) => kelvinToCelsius(rowData.main.feels_like)}
            flexGrow={1}
          />
          <Column
            width={120}
            label="Temp"
            dataKey="temp"
            cellDataGetter={({rowData}) => kelvinToCelsius(rowData.main.temp)}
            flexGrow={1}
          />
          <Column
            width={120}
            label="Temp min"
            dataKey="temp_min"
            cellDataGetter={({rowData}) => kelvinToCelsius(rowData.main.temp_min)}
            flexGrow={1}
          />
          <Column
            width={120}
            label="Temp max"
            dataKey="temp_max"
            cellDataGetter={({rowData}) => kelvinToCelsius(rowData.main.temp_max)}
            flexGrow={1}
          />
        </Table>
      )}
    </AutoSizer>
  )
}

WeatherTable.propTypes = {
  onSelect: PropTypes.func,
  weather: PropTypes.shape({
    list: PropTypes.arrayOf(
      PropTypes.shape({
        clouds: PropTypes.shape({all: PropTypes.number}),
        dt: PropTypes.number,
        dt_txt: PropTypes.string,
        visibility: PropTypes.number,
        weather: PropTypes.arrayOf(
          PropTypes.shape({
            id: PropTypes.number,
            main: PropTypes.string,
            description: PropTypes.string,
            icon: PropTypes.string,
          }),
        ),
        wind: PropTypes.shape({
          speed: PropTypes.number,
          deg: PropTypes.number,
        }),
        main: PropTypes.shape({
          feels_like: PropTypes.number,
          grnd_level: PropTypes.number,
          humidity: PropTypes.number,
          pressure: PropTypes.number,
          sea_level: PropTypes.number,
          temp: PropTypes.number,
          temp_kf: PropTypes.number,
          temp_max: PropTypes.number,
          temp_min: PropTypes.number,
        }).isRequired,
      }),
    ),
  }).isRequired,
}

export default WeatherTable
