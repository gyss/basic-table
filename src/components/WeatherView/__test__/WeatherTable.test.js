/** @format */

import * as React from 'react'
import {render, screen, fireEvent} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import WeatherTable from '../WeatherTable'
import weatherFixture from '../../../fixtures/weatherFixture'

describe('<WeatherTable>', () => {
  let props

  beforeEach(() => {
    props = {
      onSelect: jest.fn(),
      weather: weatherFixture,
    }
  })

  test('should render', () => {
    const {asFragment} = render(<WeatherTable {...props} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
