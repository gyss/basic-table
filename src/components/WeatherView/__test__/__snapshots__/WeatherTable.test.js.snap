// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherTable> should render 1`] = `
<DocumentFragment>
  <div
    style="overflow: visible; height: 0px; width: 0px;"
  >
    <div
      aria-colcount="13"
      aria-rowcount="40"
      class="ReactVirtualized__Table"
      role="grid"
    >
      <div
        class="ReactVirtualized__Table__headerRow headerRow"
        role="row"
        style="height: 40px; overflow: hidden; padding-right: 0px; width: 1570px;"
      >
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 200px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Date"
          >
            Date
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 110px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Visibility"
          >
            Visibility
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 100px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Wind"
          >
            Wind
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 100px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Clouds"
          >
            Clouds
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 100px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Rain"
          >
            Rain
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 140px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Grnd Level"
          >
            Grnd Level
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 120px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Humidity"
          >
            Humidity
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 120px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Pressure"
          >
            Pressure
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 120px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Sea Level"
          >
            Sea Level
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 120px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Feels like"
          >
            Feels like
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 120px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Temp"
          >
            Temp
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 120px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Temp min"
          >
            Temp min
          </span>
        </div>
        <div
          class="ReactVirtualized__Table__headerColumn"
          role="columnheader"
          style="flex: 1 1 120px;"
        >
          <span
            class="ReactVirtualized__Table__headerTruncatedText"
            title="Temp max"
          >
            Temp max
          </span>
        </div>
      </div>
      <div
        aria-label="grid"
        class="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
        role="rowgroup"
        style="box-sizing: border-box; direction: ltr; height: -40px; position: relative; width: 1570px; will-change: transform; overflow-x: hidden; overflow-y: auto;"
        tabindex="0"
      />
    </div>
  </div>
  <div
    class="resize-triggers"
  >
    <div
      class="expand-trigger"
    >
      <div
        style="width: 1px; height: 1px;"
      />
    </div>
    <div
      class="contract-trigger"
    />
  </div>
</DocumentFragment>
`;
